src/FffGcodeWriter.cpp

```diff
@@ -1660,57 +1732,86 @@ bool FffGcodeWriter::addSupportInfillToGCode(const SliceDataStorage& storage, La
-            gcode_layer.setIsInside(false); // going to print stuff outside print object, i.e. support
-            gcode_layer.addPolygonsByOptimizer(support_polygons, &gcode_layer.configs_storage.support_infill_config);
-            gcode_layer.addLinesByOptimizer(support_lines, &gcode_layer.configs_storage.support_infill_config, (support_pattern == EFillMethod::ZIG_ZAG)? SpaceFillType::PolyLines : SpaceFillType::Lines);
-            added = true;
+            if (support_infill_part.gradual_infill_areas_per_combine_per_density[density_idx].empty())
+            {
+                continue;
+            }
+            const Polygons& support_area = support_infill_part.gradual_infill_areas_per_combine_per_density[density_idx][0];
+
+            unsigned int density_factor = 2 << density_idx; // == pow(2, density_idx + 1)
+            int support_line_distance_here = support_line_distance * density_factor; // the highest density infill combines with the next to create a grid with density_factor 1
+            int support_shift = support_line_distance_here / 2;
+            if (density_idx == support_infill_part.gradual_infill_areas_per_combine_per_density.size() - 1)
+            {
+                support_line_distance_here /= 2;
+            }
+
+            int offset_from_outline = 0;
+            int support_infill_overlap = 0; // support infill should not be expanded outward
+
+            if (support_pattern == EFillMethod::GRID || support_pattern == EFillMethod::TRIANGLES || support_pattern == EFillMethod::CONCENTRIC)
+            {
+                support_infill_overlap = infill_extr_here.getSettingInMicrons("infill_overlap_mm"); // support lines area should be expanded outward to overlap with the boundary polygon
+            }
+
+            int extra_infill_shift = 0;
+            extra_infill_shift = support_shift;
+
+            bool use_endpieces = true;
+            Polygons* perimeter_gaps = nullptr;
+            double fill_angle = 0;
+            Infill infill_comp(support_pattern, support_area, offset_from_outline, support_line_width, support_line_distance_here, support_infill_overlap, fill_angle, z, extra_infill_shift, perimeter_gaps, infill_extr.getSettingBoolean("support_connect_zigzags"), use_endpieces);
+            Polygons support_polygons;
+            Polygons support_lines;
+            infill_comp.generate(support_polygons, support_lines);
+            if (support_lines.size() > 0 || support_polygons.size() > 0)
+            {
+                setExtruder_addPrime(storage, gcode_layer, layer_nr, infill_extruder_nr_here); // only switch extruder if we're sure we're going to switch
+                gcode_layer.setIsInside(false); // going to print stuff outside print object, i.e. support
+                gcode_layer.addPolygonsByOptimizer(support_polygons, &gcode_layer.configs_storage.support_infill_config);
```

Comment:
You're still adding the different densities one-by-one, rather than all at the same time. This code should be outside of the loop which loops over the different `density_idx`s.